File: frontend/lib/models/ticket.dart
// lib/models/ticket.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:flutter/material.dart';
import 'agent.dart';
import 'sla.dart';

part 'ticket.g.dart';

@JsonSerializable(explicitToJson: true)
class Ticket {
  @JsonKey(name: '_id') // Map MongoDB _id to id
  final String id;
  final String title;
  final String description;
  final String status;
  final int priority;
  final String category;
  final DateTime dueDate;
  final double estimatedHours;

  @JsonKey(fromJson: _agentFromJson, toJson: _agentToJson)
  final Agent? assignedTo; // Agent who the ticket is assigned to

  @JsonKey(name: 'createdBy', fromJson: _createdByFromJson)
  final String createdBy;

  final String department;

  @JsonKey(defaultValue: [])
  final List<String> requiredSkills;

  @JsonKey(fromJson: _slaFromJson, toJson: _slaToJson)
  final TicketSLA? sla;

  @JsonKey(defaultValue: 0)
  final int escalationLevel;

  @JsonKey(fromJson: _historyFromJson)
  final List<TicketHistory> history;

  final DateTime? firstResponseTime;
  final DateTime? resolvedAt;
  final int? resolutionTime;
  final DateTime createdAt;
  final DateTime updatedAt;

  Ticket({
    required this.id,
    required this.title,
    required this.description,
    required this.status,
    required this.priority,
    required this.category,
    required this.dueDate,
    required this.estimatedHours,
    this.assignedTo,
    required this.createdBy,
    required this.department,
    required this.requiredSkills,
    this.sla,
    this.escalationLevel = 0,
    required this.history,
    this.firstResponseTime,
    this.resolvedAt,
    this.resolutionTime,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Ticket.fromJson(Map<String, dynamic> json) {
    // Copy json map to avoid modifying the original
    final jsonCopy = Map<String, dynamic>.from(json);

    // Handle _id if exists
    if (jsonCopy.containsKey('_id') && !jsonCopy.containsKey('id')) {
      jsonCopy['id'] = jsonCopy['_id'];
    }

    try {
      return _$TicketFromJson(jsonCopy);
    } catch (e) {
      print('Error in Ticket.fromJson: $e');
      throw e;
    }
  }

  Map<String, dynamic> toJson() => _$TicketToJson(this);

  // Custom JSON conversion methods for Agent
  static Agent? _agentFromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    try {
      return Agent.fromJson(json);
    } catch (e) {
      print('Error parsing Agent: $e');
      return null;
    }
  }

  static Map<String, dynamic>? _agentToJson(Agent? agent) => agent?.toJson();

  // Parsing createdBy which could be a string ID or an object with _id
  static String _createdByFromJson(dynamic createdBy) {
    if (createdBy is String) {
      return createdBy;
    } else if (createdBy is Map<String, dynamic> &&
        createdBy.containsKey('_id')) {
      return createdBy['_id'].toString();
    }
    throw Exception('Invalid createdBy format: $createdBy');
  }

  // Custom JSON conversion methods for SLA
  static TicketSLA? _slaFromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    try {
      return TicketSLA.fromJson(json);
    } catch (e) {
      print('Error parsing SLA: $e');
      return null;
    }
  }

  static Map<String, dynamic>? _slaToJson(TicketSLA? sla) => sla?.toJson();

  // Handle history parsing
  static List<TicketHistory> _historyFromJson(dynamic history) {
    if (history is! List) return [];

    return history.map((item) {
      try {
        if (item is Map<String, dynamic>) {
          return TicketHistory.fromJson(item);
        }
        throw Exception('History item is not a Map: $item');
      } catch (e) {
        print('Error parsing history item: $e');
        // Return a placeholder history item
        return TicketHistory(
          action: 'error',
          timestamp: DateTime.now(),
        );
      }
    }).toList();
  }

  // Getters for derived properties
  bool get isOverdue =>
      dueDate.isBefore(DateTime.now()) &&
      status != 'resolved' &&
      status != 'closed';

  bool get isEscalated => escalationLevel > 0 || status == 'escalated';

  // New getter to fix the missing isSLABreached property
  bool get isSLABreached => sla?.isBreached ?? false;

  String get statusDisplay => status
      .split('-')
      .map((word) => word[0].toUpperCase() + word.substring(1))
      .join(' ');

  String get priorityText {
    switch (priority) {
      case 1:
        return 'High';
      case 2:
        return 'Medium';
      case 3:
        return 'Low';
      default:
        return 'Unknown';
    }
  }

  // Create a copy of the ticket with updated fields
  Ticket copyWith({
    String? id,
    String? title,
    String? description,
    String? status,
    int? priority,
    String? category,
    DateTime? dueDate,
    double? estimatedHours,
    Agent? assignedTo,
    String? createdBy,
    String? department,
    List<String>? requiredSkills,
    TicketSLA? sla,
    int? escalationLevel,
    List<TicketHistory>? history,
    DateTime? firstResponseTime,
    DateTime? resolvedAt,
    int? resolutionTime,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Ticket(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      category: category ?? this.category,
      dueDate: dueDate ?? this.dueDate,
      estimatedHours: estimatedHours ?? this.estimatedHours,
      assignedTo: assignedTo ?? this.assignedTo,
      createdBy: createdBy ?? this.createdBy,
      department: department ?? this.department,
      requiredSkills: requiredSkills ?? this.requiredSkills,
      sla: sla ?? this.sla,
      escalationLevel: escalationLevel ?? this.escalationLevel,
      history: history ?? this.history,
      firstResponseTime: firstResponseTime ?? this.firstResponseTime,
      resolvedAt: resolvedAt ?? this.resolvedAt,
      resolutionTime: resolutionTime ?? this.resolutionTime,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

@JsonSerializable()
class TicketHistory {
  final String action;
  final String? performedBy;
  final DateTime timestamp;
  final Map<String, dynamic>? details;

  TicketHistory({
    required this.action,
    this.performedBy,
    required this.timestamp,
    this.details,
  });

  factory TicketHistory.fromJson(Map<String, dynamic> json) {
    try {
      // Create a copy of the map to avoid modifying the original
      final jsonCopy = Map<String, dynamic>.from(json);

      // Handle MongoDB _id
      if (jsonCopy.containsKey('_id') && !jsonCopy.containsKey('id')) {
        jsonCopy['id'] = jsonCopy['_id'];
      }

      // Map the timestamp if it's missing
      if (!jsonCopy.containsKey('timestamp') && jsonCopy.containsKey('date')) {
        jsonCopy['timestamp'] = jsonCopy['date'];
      }

      return _$TicketHistoryFromJson(jsonCopy);
    } catch (e) {
      print('Error in TicketHistory.fromJson: $e');
      return TicketHistory(
        action: 'error',
        timestamp: DateTime.now(),
        details: {'error': e.toString()},
      );
    }
  }

  Map<String, dynamic> toJson() => _$TicketHistoryToJson(this);
}

File: backend/ticket-system-backend/src/models/Ticket.js
// src/models/Ticket.js
const mongoose = require("mongoose");

const ticketHistorySchema = new mongoose.Schema({
	action: {
		type: String,
		enum: [
			"created",
			"updated",
			"assigned",
			"reassigned",
			"escalated",
			"handover",
			"resolved",
			"closed",
			"sla_breach",
			"status_updated",
			"requeued",
		],
		required: true,
	},
	performedBy: {
		type: mongoose.Schema.Types.ObjectId,
		ref: "User",
	},
	timestamp: {
		type: Date,
		default: Date.now,
	},
	details: {
		type: mongoose.Schema.Types.Mixed,
	},
});

const slaSchema = new mongoose.Schema({
	responseTime: {
		deadline: Date,
		met: {
			type: Boolean,
			default: false,
		},
	},
	resolutionTime: {
		deadline: Date,
		met: {
			type: Boolean,
			default: false,
		},
	},
	escalationLevel: {
		type: Number,
		default: 0,
	},
	lastEscalation: Date,
});

const ticketSchema = new mongoose.Schema(
	{
		title: {
			type: String,
			required: [true, "Please provide ticket title"],
			trim: true,
		},
		description: {
			type: String,
			required: [true, "Please provide ticket description"],
			trim: true,
		},
		status: {
			type: String,
			enum: [
				"queued",
				"assigned",
				"in-progress",
				"resolved",
				"closed",
				"escalated",
			],
			default: "queued",
		},
		priority: {
			type: Number,
			required: true,
			enum: [1, 2, 3], // 1: High, 2: Medium, 3: Low
			default: 2,
		},
		category: {
			type: String,
			enum: ["technical", "billing", "general", "urgent"],
			default: "general",
		},
		dueDate: {
			type: Date,
			required: true,
		},
		estimatedHours: {
			type: Number,
			required: true,
			default: 1,
		},
		assignedTo: {
			type: mongoose.Schema.Types.ObjectId,
			ref: "Agent",
			default: null,
		},
		createdBy: {
			type: mongoose.Schema.Types.ObjectId,
			ref: "User",
			required: true,
		},
		department: {
			type: String,
			required: true,
			default: "general",
		},
		requiredSkills: [String],
		sla: {
			type: slaSchema,
			default: () => ({}),
		},
		escalationLevel: {
			type: Number,
			default: 0,
		},
		history: [ticketHistorySchema],
		firstResponseTime: Date,
		resolvedAt: Date,
		resolutionTime: Number, // in minutes
	},
	{
		timestamps: true,
	}
);

// Indexes
ticketSchema.index({ status: 1, priority: 1, createdAt: -1 });
ticketSchema.index({ assignedTo: 1, status: 1 });
ticketSchema.index({ createdBy: 1, status: 1 });
ticketSchema.index({ dueDate: 1 });
ticketSchema.index({ priority: 1, dueDate: 1 });
ticketSchema.index({ "sla.responseTime.deadline": 1 });
ticketSchema.index({ "sla.resolutionTime.deadline": 1 });

// Methods
ticketSchema.methods.isOverdue = function () {
	return (
		this.dueDate < new Date() &&
		this.status !== "resolved" &&
		this.status !== "closed"
	);
};

ticketSchema.methods.needsEscalation = function () {
	return (
		this.sla &&
		((this.sla.responseTime &&
			!this.sla.responseTime.met &&
			new Date() > this.sla.responseTime.deadline) ||
			(this.sla.resolutionTime &&
				!this.sla.resolutionTime.met &&
				new Date() > this.sla.resolutionTime.deadline))
	);
};

// Pre-save hook for SLA initialization
ticketSchema.pre("save", async function (next) {
	if (this.isNew && !this.sla?.responseTime?.deadline) {
		try {
			// Instead of requiring SLAService directly, calculate deadlines here
			const config = (await SLAConfig.findOne({
				priority: this.priority,
				category: this.category,
			})) || {
				responseTime: 60,
				resolutionTime: 480,
			};

			const now = new Date();
			this.sla = {
				responseTime: {
					deadline: new Date(now.getTime() + config.responseTime * 60000),
					met: false,
				},
				resolutionTime: {
					deadline: new Date(now.getTime() + config.resolutionTime * 60000),
					met: false,
				},
			};
		} catch (error) {
			console.error("Error initializing SLA for ticket:", error);
		}
	}
	next();
});

module.exports = mongoose.model("Ticket", ticketSchema);
